nengo/connection.py

```diff
@@ -319,6 +319,28 @@ def repeated_inds(x):
         return super().coerce(conn, transform)
 
 
+class ConnectionInitialValueParam(DistOrArrayParam):
+    """Connection-specific validation for initial values."""
+
+    def coerce(self, conn, value):
+        value = super().coerce(conn, value)
+
+        if isinstance(value, np.ndarray):
+            if isinstance(conn.post_obj, Ensemble) and conn.solver.weights:
+                n_synapses = conn.post_obj.n_neurons
+            else:
+                n_synapses = conn.size_out
```

Comment:
Does this code path have good testing coverage? I wonder if this logic handles all cases, given how complex the build connection function is when it comes to all the case logic with slicing and direct neuron connections. An easy way to enforce testing coverage here would be to change the default to `initial_value=0` but that raises the error I mention below.

Reply:
I think it's actually the other part of the path (using a weight solver) that isn't well tested. This is tested in `test_connection.py:test_bad_initial_values`, and that case isn't covered there. You're right that we don't test all possible types of connections, but having `initial_value=0` as the default wouldn't check this either because then `value.ndim > 0` would be false and `n_synapses` wouldn't even get checked. If we wanted to test this, we'd have to explicitly test all those connection types, either by adding to `test_initial_value` or by setting initial values as vectors in some of the tests that test those other connection types. Personally, I'm not too worried. This choice is based of an understanding of how the connection builder works, and the number of synapses always is equal to the connection `size_out`.

Reply:
This check is now gone. Since we've moved `initial_output` to `Synapse`, we can only validate the shape once we actually build.